File: uni_modules/utssdk/app-ios/src/LinphoneBridge.swift
import Foundation
import AVFoundation

// TODO: Ensure the Linlin frameworks are linked in the consuming Xcode target (see app-ios/Frameworks).

@objc public protocol LinphoneBridgeListener: AnyObject {
    func linphoneBridge(_ bridge: LinphoneBridge, didReceiveEvent type: String, payload: NSDictionary)
    func linphoneBridge(_ bridge: LinphoneBridge, didEncounterError error: NSError)
}

@objcMembers
public final class LinphoneBridge: NSObject {
    public static let shared = LinphoneBridge()

    private let manager = LinphoneManager.shared
    private let audioRouter = AudioRouter()
    private let callbackQueue = DispatchQueue.main

    private var eventCallback: ((String, NSDictionary) -> Void)?

    public weak var listener: LinphoneBridgeListener?

    private override init() {
        super.init()
        manager.delegate = self
    }

    public func initialize(with config: NSDictionary?, listener: LinphoneBridgeListener? = nil) {
        if let listener = listener {
            self.listener = listener
        }

        requestMicrophonePermission { [weak self] granted in
            guard let self = self else { return }

            guard granted else {
                self.emitEvent(type: "permission", payload: [
                    "resource": "microphone",
                    "status": "denied"
                ])

                let error = NSError(
                    domain: "LinphoneBridge.Permission",
                    code: 1,
                    userInfo: [NSLocalizedDescriptionKey: "Microphone permission denied"]
                )
                self.emitError("permission", error: error)
                return
            }

            let configuration = config as? [String: Any] ?? [:]
            self.emitEvent(type: "permission", payload: [
                "resource": "microphone",
                "status": "granted"
            ])
            self.manager.initialize(with: configuration)

            self.emitEvent(type: "registration", payload: [
                "state": LinphoneManager.RegistrationState.idle.rawValue,
                "message": "Initialized"
            ])
        }
    }

    public func setEventCallback(_ callback: @escaping (String, NSDictionary) -> Void) {
        eventCallback = callback
    }

    public func clearEventCallback() {
        eventCallback = nil
    }

    public func register() {
        manager.register()
    }

    public func unregister() {
        manager.unregister()
    }

    public func callDial(number: String) {
        manager.dial(number: number)
    }

    public func callHangup() {
        manager.hangup()
    }

    public func callAnswer() {
        manager.answer()
    }

    public func sendDtmf(tone: String) {
        manager.sendDtmf(tone: tone)
    }

    public func messageSend(to recipient: String, text: String) {
        manager.sendMessage(to: recipient, text: text)
    }

    public func audioSetRoute(route: String) {
        let selectedRoute = AudioRoute(rawValue: route.lowercased()) ?? .system

        do {
            try audioRouter.setRoute(selectedRoute)
            manager.updateAudioRoute(selectedRoute)
            emitEvent(type: "audioRoute", payload: [
                "route": selectedRoute.rawValue
            ])
        } catch {
            emitError("audioRoute", error: error)
        }
    }

    private func requestMicrophonePermission(completion: @escaping (Bool) -> Void) {
        let session = AVAudioSession.sharedInstance()

        switch session.recordPermission {
        case .granted:
            completion(true)
        case .denied:
            completion(false)
        case .undetermined:
            session.requestRecordPermission { [weak self] granted in
                guard let self = self else { return }
                self.callbackQueue.async {
                    completion(granted)
                }
            }
        @unknown default:
            completion(false)
        }
    }

    private func emitEvent(type: String, payload: [String: Any]) {
        callbackQueue.async { [weak self] in
            guard let self = self else { return }
            let dictionary = payload as NSDictionary
            self.listener?.linphoneBridge(self, didReceiveEvent: type, payload: dictionary)
            self.eventCallback?(type, dictionary)
        }
    }

    private func emitError(_ type: String, error: Error) {
        let nsError = error as NSError
        callbackQueue.async { [weak self] in
            guard let self = self else { return }
            self.listener?.linphoneBridge(self, didEncounterError: nsError)
        }
        emitEvent(type: "error", payload: [
            "category": type,
            "message": error.localizedDescription
        ])
    }
}

extension LinphoneBridge: LinphoneManagerDelegate {
    public func linphoneManager(_ manager: LinphoneManager, didUpdateRegistration state: LinphoneManager.RegistrationState, message: String?) {
        var payload: [String: Any] = ["state": state.rawValue]
        if let message = message {
            payload["message"] = message
        }
        emitEvent(type: "registration", payload: payload)
    }

    public func linphoneManager(_ manager: LinphoneManager, didUpdateCall state: LinphoneManager.CallState, info: [String: Any]) {
        var payload = info
        payload["state"] = state.rawValue
        emitEvent(type: "call", payload: payload)
    }

    public func linphoneManager(_ manager: LinphoneManager, didReceiveMessage payload: [String: Any]) {
        emitEvent(type: "message", payload: payload)
    }

    public func linphoneManager(_ manager: LinphoneManager, didChangeAudioRoute route: AudioRoute) {
        emitEvent(type: "audioRoute", payload: [
            "route": route.rawValue
        ])
    }

    public func linphoneManager(_ manager: LinphoneManager, didFailWith error: Error, category: String) {
        emitError(category, error: error)
    }
}

File: uni_modules/utssdk/app-android/src/main/java/com/utssdk/linphone/LinphoneBridge.kt
package com.utssdk.linphone

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.os.Handler
import android.os.Looper
import androidx.core.content.ContextCompat
import kotlin.jvm.functions.Function2

object LinphoneBridge : LinphoneManager.Callback {

    interface LinphoneEventListener {
        fun onEvent(type: String, payload: Map<String, Any?>)
        fun onError(category: String, message: String)
    }

    private val manager = LinphoneManager
    private val audioRouter = AudioRouter()


    private var applicationContext: Context? = null
    private var listener: LinphoneEventListener? = null
    private var eventCallback: Function2<String, Map<String, Any?>, Unit>? = null

    @JvmStatic
    @JvmOverloads
    fun initialize(context: Context, config: Map<String, Any?>? = emptyMap(), listener: LinphoneEventListener? = null) {
        applicationContext = context.applicationContext
        listener?.let { this.listener = it }
        manager.setCallback(this)

        if (!ensureMicrophonePermission()) {
            emitEvent(
                type = "permission",
                payload = mapOf(
                    "resource" to "microphone",
                    "status" to "denied"
                )
            )
            emitError("permission", IllegalStateException("Microphone permission denied"))
            return
        }

        val safeContext = requireNotNull(applicationContext) {
            "Application context must be set before calling initialize"
        }

        val configMap = config ?: emptyMap()
        val serviceConfig = LinphoneConfig.from(configMap)

        manager.initialize(safeContext, configMap)
        try {
            LinphoneService.start(safeContext, serviceConfig)
        } catch (throwable: Throwable) {
            emitError("service", throwable)
        }

        emitEvent(
            type = "permission",
            payload = mapOf(
                "resource" to "microphone",
                "status" to "granted"
            )
        )
        emitEvent(
            type = "registration",
            payload = mapOf(
                "state" to LinphoneManager.RegistrationState.IDLE.wireName,
                "message" to "Initialized"
            )
        )
        }

        @JvmStatic
        fun setEventListener(listener: LinphoneEventListener?) {
        this.listener = listener
        }

        @JvmStatic
        fun setEventCallback(callback: Function2<String, Map<String, Any?>, Unit>?) {
        this.eventCallback = callback
        }

        @JvmStatic
        fun clearEventCallback() {
        this.eventCallback = null
        }

        @JvmStatic

        manager.register()
    }

    @JvmStatic
    fun unregister() {
        manager.unregister()
    }

    @JvmStatic
    fun callDial(number: String) {
        manager.dial(number)
    }

    @JvmStatic
    fun callHangup() {
        manager.hangup()
    }

    @JvmStatic
    fun callAnswer() {
        manager.answer()
    }

    @JvmStatic
    fun sendDtmf(tone: String) {
        manager.sendDtmf(tone)
    }

    @JvmStatic
    fun messageSend(recipient: String, text: String) {
        manager.sendMessage(recipient, text)
    }

    @JvmStatic
    fun audioSetRoute(route: String) {
        val context = applicationContext
        if (context == null) {
            emitError("audioRoute", IllegalStateException("Context is not available"))
            return
        }

        val targetRoute = AudioRouter.Route.fromValue(route)
        val applied = audioRouter.applyRoute(context, targetRoute)
        if (applied) {
            manager.updateAudioRoute(targetRoute)
            emitEvent("audioRoute", mapOf("route" to targetRoute.wireName))
        } else {
            emitError("audioRoute", IllegalStateException("Failed to apply audio route"))
        }
    }

    private fun ensureMicrophonePermission(): Boolean {
        val context = applicationContext ?: return false
        val granted = ContextCompat.checkSelfPermission(
            context,
            Manifest.permission.RECORD_AUDIO
        ) == PackageManager.PERMISSION_GRANTED

        if (!granted) {
            // TODO: Integrate runtime permission request flow once host Activity is wired.
        }
        return granted
    }

    private fun emitEvent(type: String, payload: Map<String, Any?>) {
        mainHandler.post {
            listener?.onEvent(type, payload)
            eventCallback?.invoke(type, payload)
        }
    }

    private fun emitError(category: String, throwable: Throwable) {
        val message = throwable.message ?: category
        mainHandler.post {
            listener?.onError(category, message)
        }
        emitEvent(
            type = "error",
            payload = mapOf(
                "category" to category,
                "message" to message
            )
        )
    }

    override fun onRegistrationStateChanged(state: LinphoneManager.RegistrationState, message: String?) {
        val payload = mutableMapOf<String, Any?>("state" to state.wireName)
        if (!message.isNullOrEmpty()) {
            payload["message"] = message
        }
        emitEvent("registration", payload)
    }

    override fun onCallStateChanged(state: LinphoneManager.CallState, info: Map<String, Any?>) {
        val payload = info.toMutableMap()
        payload["state"] = state.wireName
        emitEvent("call", payload)
    }

    override fun onMessage(payload: Map<String, Any?>) {
        emitEvent("message", payload)
    }

    override fun onAudioRoute(route: AudioRouter.Route) {
        emitEvent("audioRoute", mapOf("route" to route.wireName))
    }

    override fun onError(category: String, throwable: Throwable) {
        emitError(category, throwable)
    }
}
